### ISSUE TITLE
Definition of subdomain_id() on serial triangulation

### ISSUE THREAD

**Issue by kronbichler** [dealii/dealii#8801]
While looking into #8772, I saw that `TriaAccessor::is_locally_owned()` etc all perform a `dynamic_cast` to see whether the triangulation is actually a more derived (parallel) class. This is reflected in the documentation https://www.dealii.org/current/doxygen/deal.II/classCellAccessor.html#a87a3ff6217f3b5b4537f6a101560f493 where we state that this is always true for serial triangulation. I find it confusing that we need the dynamic cast and why we do not have `cell->subdomain_id() == tria.locally_owned_subdomain()`. (I did a quick check on a serial case and realized that we have assigned 0 to the former and `-1` to the latter.)

I vaguely remember that we once mis-used the subdomain id also for other purposes, but thought we had stopped. Is there a reason to not simply have the two in a consistent state? Or does somebody remember what the design was?

As an alternative to the dynamic cast, I think the check for a serial triangulation would be cheaper if we simply asked `cell->tria->locally_owned_subdomain() == numbers::invalid_subdomain_id`. (`dynamic_cast` has to do some string operations to check for the actual class type which are non-negligible.)

**Comment by bangerth** [comment]
It used to be that `subdomain_id` was a user-settable field, and that's how we treated it. For example, that's how step-18 was implemented for the first ten years: Via a regular (=sequential) triangulation on which we somehow set the `subdomain_id`s. It was only when @davydden wrote the `parallel::shared::Triangulation` that we gained a way to make setting the `subdomain_id` something that's automatic in any kind of sense -- before, you could have come up with whatever scheme you liked and put it into the subdomain ids of a triangulation. This field was in essence another `user_index` that was interpreted in some particular way.

I feel like we don't really want to allow this any more. I wouldn't mind if we initialized this field in some way that's convenient, and document that it's for the exclusive use of the parallel triangulations. That would make it possible to do what you suggest. If we felt strongly about it, we could put assertions into the triangulation (or the `cell->subdomain_id()` function) to ensure that the field can only be set to anything if it's not a sequential triangulation.

**Comment by peterrum** [comment]
@bangerth I would like to point out that `cell->set_subdomain_id()` is currently used during partitioning of serial triangulations (see `GridTools` and `parallel::fullydistributed::Utilities`).

**Comment by bangerth** [comment]
Oh, that's unfortunate.

So what concretely *is* @kronbichler 's suggestion then, if this field is currently used for serial triangulations as well?

**Comment by kronbichler** [comment]
I did not have a suggestion at the point when opening the issue - I was wondering if we could streamline the use within the library, i.e., have `cell->subdomain_id()` matching with `Triangulation::locally_owned_subdomain()`. Furthermore, I wanted to know about use cases others know before I searched the whole library. The use cases suggest that it is a bit more complicated than I thought, and as @bangerth and @peterrum suggest we cannot prohibit `set_subdomain_id()` on a serial triangulation. So I would suggest the following now:
1. I replace the `dynamic_cast` to a parallel triangulation by checking `locally_owned_subdomain()==numbers::invalid_subdomain_id` instead.
2. As a possibility, we could allow setting the subdomain id of a serial triangulation, so it can be used for constructing another kind of triangulation, but without allowing to query the ghost/locally owned state any more. This would break code that pre-dates the `parallel::shared::Triangulation` concept. Not sure if it is worth it, though.
